### PR TITLE
PHPC-2005: Support 'let' option for multiple CRUD commands

### PR DESCRIPTION
**Pull request by jmikola**
https://jira.mongodb.org/browse/PHPC-2005

### CODE REVIEW

File: tests/bulk/bulkwrite-ctor-let_error-001.phpt [hunk 1]
```diff
@@ -0,0 +1,18 @@
+--TEST--
+MongoDB\Driver\BulkWrite::__construct(): let option
+--FILE--
+<?php
+
+require_once __DIR__ . '/../utils/basic.inc';
+
+echo throws(function() {
```
Comment (aleksandr-rudo) [inline tests/bulk/bulkwrite-ctor-let_error-001.phpt:8]: May be it would be more safe to add a various of incorrect data tests, for example NULL value, which sometime produces borderline cases?

File: tests/bulk/bulkwrite-ctor-let-001.phpt [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+var_dump($cursor->toArray());
+
+?>
+===DONE===
+<?php exit(0); ?>
+--EXPECTF--
+insert does not include let option
+delete included let: {"id":1,"x":"foo"}
+update included let: {"id":1,"x":"foo"}
+find does not include let option
```
[the remark this answers is not in the document]
Comment (aleksandr-rudo) [reply tests/bulk/bulkwrite-ctor-let-001.phpt:58]: As far as I get the idea, this test is designed for BulkWrite constructor and that's why it includes both delete and update operations. But may be in such case it would be more flexible to arrange constructor tests not for single argument?

File: tests/bulk/bulkwrite-ctor-let_error-001.phpt (hunk 1 above)
Comment (jmikola) [reply tests/bulk/bulkwrite-ctor-let_error-001.phpt]: This is far less likely to happen in the extension since we don't use things like `isset()` and examine the zval type directly. For example:

```c
if (Z_TYPE_P(value) != IS_OBJECT && Z_TYPE_P(value) != IS_ARRAY) {
```

That said, there's no reason we can't test more invalid types. In older PHP versions, some types were reported inconsistently (e.g. "integer" vs. "int"), but I don't think that's an issue with our current version requirement of PHP 7.2+.

I added some additional test cases.

File: tests/bulk/bulkwrite-ctor-let-001.phpt (hunk 2 above)
[the remark this answers is not in the document]
Comment (jmikola) [reply tests/bulk/bulkwrite-ctor-let-001.phpt:58]: >  I meant tests arrangement looks like they a separated in groups of actions: delete, insert, update. So constructor tests (as you added prefix "ctor" for new test) can be arranged in separate group, which can verify not only "let" argument, but may be some more actions or arguments.

Following up on your Slack comment. The convention for naming tests is generally `class-method-###.phpt`, and there's some variation for `_error` tests as you've seen. For things that don't have methods, we typically just fill in the name of the operation instead of "method", such as `debug` for testing the `get_debug_info` handler (via `var_dump()`) or `serialize` for testing how the object behaves with `serialize()`.

Since some of our constructors support _many_ options, it's not feasible to stick to `class-ctor-###.phpt` for all of them, as the filenames wouldn't be very useful and we'd just end up with lots and lots of tests. The tests for the Manager class are a good example of this where we have `manager-ctor-optionname-###.phpt` and that's what I've decided to borrow here. Using "ctor" is just personal preference since "__construct` would be more verbose and both convey the same information.

If/when we introduce a `bypassDocumentValidation` test, we can certainly do so in a new `bulkwrite-ctor-bypassdocumentvalidation-001.phpt` file. And seeing the existing `let` option test being added in this PR would likely serve as an example for the author to do so down the line.

I'll also point out that most of the actual method tests (e.g. `bulkwrite-update-*`) only use a numeric sequence, as demonstrated by the `BulkWrite::update()` test for `hint` I linked above. There typically aren't very many tests for individual methods are there are for constructors, which is why we've not bothered to create names like `bulkwrite-update-hint-001.phpt` -- although we certainly could. In the case of `let`, this was the very first `bulkwrite-ctor` test so I decided to adopt the style we employed for Manager.
